compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp

```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
+
+  // CHeck that the LHS operand is the same. Can be relaxed in the future.
+  if (aLinalgOp.getDpsInputOperand(0)->get() !=
+      bLinalgOp.getDpsInputOperand(0)->get()) {
+    return false;
+  }
+
+  auto checkSameRankAndElementType = [](Value aVal, Value bVal) {
+    auto aType = dyn_cast<ShapedType>(aVal.getType());
+    auto bType = dyn_cast<ShapedType>(bVal.getType());
+    return aType && bType && aType.getRank() == bType.getRank() &&
+           aType.getElementType() == bType.getElementType();
   };
-  auto markEquivalent = [&](Value v1, Value v2) { equivalentValues[v1] = v2; };
-  return OperationEquivalence::isEquivalentTo(
-      lhsOp, rhsOp, checkEquivalent, markEquivalent,
-      /*flags=*/OperationEquivalence::IgnoreLocations);
-}
+  // Check that the RHS rank and element type are the same. We dont check the
+  // type cause we allow RHS to be transposes.
+  if (!checkSameRankAndElementType(aLinalgOp.getDpsInputOperand(1)->get(),
+                                   bLinalgOp.getDpsInputOperand(1)->get())) {
+    return false;
+  }
 
-/// Check that an operation is a `empty -> fill -> contraction`
-static bool isEmptyFillContractionDAGRootOp(
-    linalg::LinalgOp linalgOp,
-    std::optional<linalg::LinalgOp> seedContractionOp = std::nullopt) {
-  if (!linalg::isaContractionOpInterface(linalgOp)) {
+  // Check that the output rank and element type are the same. We dont check the
+  // type cause we allow RHS to be transposes.
+  if (!checkSameRankAndElementType(aLinalgOp.getDpsInitOperand(0)->get(),
+                                   bLinalgOp.getDpsInitOperand(0)->get())) {
     return false;
   }
-  auto fillOp = linalgOp.getDpsInits()[0].getDefiningOp<linalg::FillOp>();
-  if (!fillOp) {
+
+  // Check that the iterator types are the same.
+  if (aLinalgOp.getIteratorTypesArray() != bLinalgOp.getIteratorTypesArray()) {
     return false;
   }
-  // For convenience check that the fill value is 0. This is not
-  // a necessity, but easier to handle the rewrite this way.
-  if (!matchPattern(fillOp.getDpsInputOperand(0)->get(), m_AnyZeroFloat()) &&
-      !matchPattern(fillOp.getDpsInputOperand(0)->get(), m_Zero())) {
+
+  // Check region equivalence.
+  if (!OperationEquivalence::isRegionEquivalentTo(
+          &aLinalgOp->getRegion(0), &bLinalgOp->getRegion(0),
+          OperationEquivalence::IgnoreLocations)) {
     return false;
   }
-  if (!fillOp.getDpsInitOperand(0)->get().getDefiningOp<tensor::EmptyOp>()) {
+
+  return true;
+}
+
+/// Check that an operation is a `empty -> fill -> contraction`
+static bool isEquivalentContractionOp(
+    linalg::LinalgOp linalgOp,
+    std::optional<linalg::LinalgOp> seedContractionOp = std::nullopt) {
+  if (!linalg::isaContractionOpInterface(linalgOp)) {
     return false;
   }
   if (seedContractionOp) {
-    return checkOperationEquivalence(linalgOp, seedContractionOp.value());
+    return checkOperationEquivalence(seedContractionOp.value(), linalgOp);
   }
   return true;
```

Comment:
How about we inline the `checkOperationEquivalence` logic to here. This function is fairly small to me, so I think we do not need an additional function call. E.g., it can be

```cpp
if (!seedContractionOp) {
  return true;
}
// the logic of checkOperationEquivalence
```

Reply:
I want to leave it this way...  The `checkContractionEquivalence` already assumes the two ops are contractions.